resolver/cluster.go

```diff
@@ -97,43 +99,52 @@ func (r *clusterResolver) LookupBindingSecret(ctx context.Context, serviceRef co
 }
 
-func (r *clusterResolver) LookupWorkloads(ctx context.Context, workloadRef corev1.ObjectReference, selector *metav1.LabelSelector) ([]runtime.Object, error) {
-	if workloadRef.Name != "" {
-		workload, err := r.lookupWorkload(ctx, workloadRef)
+const (
+	mappingAnnotationPrefix = "projector.servicebinding.io/mapping-"
+)
+
+func (r *clusterResolver) LookupWorkloads(ctx context.Context, serviceBinding *servicebindingv1beta1.ServiceBinding) ([]runtime.Object, error) {
+	workloadRef := serviceBinding.Spec.Workload
+
+	list := &unstructured.UnstructuredList{}
+	list.SetAPIVersion(workloadRef.APIVersion)
+	list.SetKind(fmt.Sprintf("%sList", workloadRef.Kind))
+
+	var ls labels.Selector
+	if workloadRef.Selector != nil {
+		var err error
+		ls, err = metav1.LabelSelectorAsSelector(workloadRef.Selector)
 		if err != nil {
 			return nil, err
 		}
-		return []runtime.Object{workload}, nil
-	}
-	return r.lookupWorkloads(ctx, workloadRef, selector)
-}
-
-func (r *clusterResolver) lookupWorkload(ctx context.Context, workloadRef corev1.ObjectReference) (runtime.Object, error) {
-	workload := &unstructured.Unstructured{}
-	workload.SetAPIVersion(workloadRef.APIVersion)
-	workload.SetKind(workloadRef.Kind)
-	if err := r.client.Get(ctx, client.ObjectKey{Namespace: workloadRef.Namespace, Name: workloadRef.Name}, workload); err != nil {
-		return nil, err
 	}
-	return workload, nil
-}
 
-func (r *clusterResolver) lookupWorkloads(ctx context.Context, workloadRef corev1.ObjectReference, selector *metav1.LabelSelector) ([]runtime.Object, error) {
-	workloads := &unstructured.UnstructuredList{}
-	workloads.SetAPIVersion(workloadRef.APIVersion)
-	workloads.SetKind(fmt.Sprintf("%sList", workloadRef.Kind))
-	ls, err := metav1.LabelSelectorAsSelector(selector)
-	if err != nil {
+	if err := r.client.List(ctx, list, client.InNamespace(serviceBinding.Namespace)); err != nil {
 		return nil, err
 	}
-	if err := r.client.List(ctx, workloads, client.InNamespace(workloadRef.Namespace), client.MatchingLabelsSelector{Selector: ls}); err != nil {
-		return nil, err
+	workloads := []runtime.Object{}
+	for i := range list.Items {
+		workload := &list.Items[i]
+		if annotations := workload.GetAnnotations(); annotations != nil {
+			if _, ok := annotations[fmt.Sprintf("%s%s", mappingAnnotationPrefix, serviceBinding.UID)]; ok {
+				workloads = append(workloads, workload)
+				continue
+			}
+		}
+		if workloadRef.Name != "" {
+			if workload.GetName() == workloadRef.Name {
+				workloads = append(workloads, workload)
+			}
+			continue
+		}
+		if ls.Matches(labels.Set(workload.GetLabels())) {
+			workloads = append(workloads, workload)
+		}
 	}
 
-	// coerce to []runtime.Object
-	result := make([]runtime.Object, len(workloads.Items))
-	for i := range workloads.Items {
-		result[i] = &workloads.Items[i]
-	}
-	return result, nil
+	sort.Slice(workloads, func(i, j int) bool {
+		return workloads[i].(metav1.Object).GetName() < workloads[j].(metav1.Object).GetName()
+	})
```

Comment:
strictly speaking no, but for tests observing side effects the order matters. In practice the api server returns items in order, but there is no guarantee. I added this because I was seeing tests fail because the items were out of order.